tonconnect-server/TonConnect2.md

```diff
@@ -0,0 +1,533 @@
+
+
+### Decentralized vs centralized apps
+
+In the above definitions apps are considered decentralized and services considered centralized. 
+In reality, there is a spectrum of trust-minimization and each specific app or service may rely on some centralized infrastructure.
+
+Both apps and services are often non-custodial for the security and legal reasons. Instead of managing cryptographic keys that hold users’ funds they rely on wallet apps to approve and sign transactions on behalf of the users. Wallet apps themselves are often non-custodial too: they store secret keys securely on their users’ devices and never communicate them to other computers on the network.
+
+
+### User identification
+
+To minimize cross-service tracking, TON Connect uses distinct cryptographic identities for each service.
+
+User identity is tied to the [app keypair](#app-keypair).
+
+
+### User authentication
+
+Decentralized apps authenticate users within smart contracts using message sender addresses. 
+The UI (“offchain logic”) runs on behalf of the user, so there is no need for a dedicated “sign in” scheme.
+
+Centralized apps run on the servers and may store user data. They may use “sign in” functionality to authenticate the user, where instead of username/password pair we use cryptographic keypair derived from the wallet secret seed.
+
+
+### App authentication
+
+Apps are authenticated using [TON DNS](https://ton.org/docs/#/web3/dns) protocol.
+
+Decentralized apps may use `sha256("wallet")` key for the smart contract address. 
+When the wallet receives request to sign a message to a smart contract with a given **.ton** name, a valid and up-to-date DNS record would be used to authenticate that contract address with that name.
+
+**TODO:** callback/return URLs should also be specified in the TON DNS to permit their verification for dapps that cannot sign their own requests.
+
+Centralized services may also use `sha256("tonconnect.app-pk")` key to specify 32-byte [app public key](#app-keypair) used to initiate authentication requests to the wallet. The key should be stored on the online server and can be frequently rotated via TON.DNS record update.
+
+
+
+## Workflows
+
+### Connect workflow
+
+1. [App](#app) or [service](#service) prepares [ConnectRequest](#ConnectRequest) and displays a QR code and a button to let the [wallet](#wallet) connect.
+2. User scans the QR or clicks the button.
+3. Wallet shows confirmation dialog that shows the wallet address communicated to the app.
+4. Wallet stores app’s web address (if provided) in the list of connected services.
+5. Wallet sends/redirects the user back to the app with the reply: [ConnectResponse](#ConnectResponse) containing its address and optional push URL.
```

Comment:
I don't understand something. In the chat group we had a very very long conversation about this redirect. How does this redirect happens? In most mobile app use-cases, it's very difficult and fragile to redirect back to the native app that asked to connect.

Consider that the user is talking to a bot through Telegram messenger. The bot shows the frontend for the app right inside Telegram Messenger. We want in the end of the connect action to send the user back to the Telegram native app. The same if the user is browsing to a TON app through Instagram native app. How can the app frontend know what's the correct return URL? It depends on the browser where it was opened from.

The only consistent way to achieve a smooth UX in this case was to poll on an AJAX endpoint from the app frontend until the user is connected. We discussed this in length. Where did this flow disappear?

Reply:
@oleganza we're trying to implement a DEX mobile client (web app) based on this standard and we need the AJAX polling method so the flow is seamless. Can you confirm this AJAX polling method is supported by the standard?